src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs

```diff
@@ -28,20 +28,36 @@ internal sealed partial class RemoteDesignerAttributeIncrementalAnalyzer : Incre
         /// </summary>
         private readonly RemoteEndPoint _endPoint;
 
-        private readonly IPersistentStorageService _storageService;
+        /// <summary>
+        /// Keep track of the last information we reported.  We will avoid notifying the host if we recompute and these
+        /// don't change.
+        /// </summary>
+        private readonly ConcurrentDictionary<DocumentId, (string? category, VersionStamp projectVersion)> _documentToLastReportedInformation =
+            new ConcurrentDictionary<DocumentId, (string? category, VersionStamp projectVersion)>();
 
-        public RemoteDesignerAttributeIncrementalAnalyzer(Workspace workspace, RemoteEndPoint endPoint)
+        public RemoteDesignerAttributeIncrementalAnalyzer(RemoteEndPoint endPoint)
         {
             _endPoint = endPoint;
-            _storageService = workspace.Services.GetRequiredService<IPersistentStorageService>();
         }
 
-        public override Task RemoveProjectAsync(ProjectId projectId, CancellationToken cancellationToken)
+        public override async Task RemoveProjectAsync(ProjectId projectId, CancellationToken cancellationToken)
```

Comment:
@mavasani for information here.  IN the case of solution-unload, how to incremetnal analyzers hear about the 'removal'?  Thanks!

Reply:
Current implementation invokes RemoveDocumentAsync for each document in the solution: http://sourceroslyn.io/#Microsoft.CodeAnalysis.Features/SolutionCrawler/WorkCoordinator.cs,407. It seems it can certainly do batching here and just invoke RemoveProjectAsync for all projects in the solution.

Reply:
on interesting.  so there's no real point in responding to RemoveProject if i'm also responding to RemoveDocument, correct?

Reply:
We invoke RemoveProject for enqueued project items for which `solution.GetProject(id)` returns null, i.e. project has been removed and we cannot find the documents for it.

Reply:
http://sourceroslyn.io/#Microsoft.CodeAnalysis.Features/SolutionCrawler/WorkCoordinator.LowPriorityProcessor.cs,153